Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp

```diff
@@ -351,58 +351,70 @@ auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromi
     return promise;
 }
 
-void MediaStreamTrack::getPhotoCapabilities(DOMPromiseDeferred<IDLDictionary<PhotoCapabilities>>&& promise)
-{
-    auto identifier = LOGIDENTIFIER;
-
-    m_private->getPhotoCapabilities([this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)](auto&& result) mutable {
-        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)] () mutable {
-            if (!result) {
-                // https://w3c.github.io/mediacapture-image/#ref-for-dom-imagecapture-getphotocapabilities②
-                // If the data cannot be gathered for any reason (for example, the MediaStreamTrack being ended
-                // asynchronously), then reject p with a new DOMException whose name is OperationError, and
-                // abort these steps.
-                ERROR_LOG(identifier, "rejecting promise: ", result.errorMessage);
-                promise.reject(Exception { ExceptionCode::OperationError, WTFMove(result.errorMessage) });
-                return;
-            }
-            if (m_readyState != State::Live) {
-                ERROR_LOG(identifier, "rejecting promise, track has ended");
-                promise.reject(Exception { ExceptionCode::OperationError, "Track has ended"_s });
-                return;
-            }
+auto MediaStreamTrack::getPhotoCapabilities() -> Ref<PhotoCapabilitiesPromise>
+{
+    // https://w3c.github.io/mediacapture-image/#dom-imagecapture-getphotocapabilities
+    // If the readyState of track provided in the constructor is not live, return
+    // a promise rejected with a new DOMException whose name is InvalidStateError,
+    // and abort these steps.
+    if (m_ended)
+        return PhotoCapabilitiesPromise::createAndReject(Exception { ExceptionCode::InvalidStateError, "Track has ended"_s });
 
-            ALWAYS_LOG(identifier, "resolving promise");
-            promise.resolve(WTFMove(*result.capabilities));
-        });
+    PhotoCapabilitiesPromise::Producer producer;
+    Ref<PhotoCapabilitiesPromise> promise = producer;
```

Comment:
you can do `producer.promise()` now to not have the style checker complain

Reply:
You don't need the producer here. 

Return the PhotoCapabilitiesPromise in the whenSettled call.